libcobblersignatures/__init__.py

```diff
@@ -2,3 +2,221 @@
+        Convert the internal data to a JSON. Only for internal usage.
+
+        :param sort_keys: If the keys of the dictionary should be sorted to be more human readable.
+        :param indent: If this is something other then ``None`` then the JSON will be pretty printed.
+        """
+        value = {}
+        for breed in self.osbreeds:
+            value[breed.name] = breed.encode()
+        return json.dumps({self._rootkey: value}, sort_keys=sort_keys, indent=indent)
+
+    def exportsignatures(self, export_type: ExportTypes, target: str = "", sort_keys: bool = False,
+                         indent: Union[None, int] = None):
+        """
+        This is the main export function.
+
+        :param export_type: One of the values from the :class:`ExportTypes`.
+        :param target: This is only required when using this for a file based export. Otherwise this can be skipped.
+        :param sort_keys: If the keys of the dictionary should be sorted to be more human readable.
+        :param indent: If this is something other then ``None`` then the JSON will be pretty printed.
+        :raises ValueError: When the :class:`ExportTypes` is not implemented or not known.
+        :raises TypeError: When one of the arguments has the wrong type.
+        """
+        if not isinstance(target, str):
+            raise TypeError("target needs to be of type str!")
+        if not isinstance(sort_keys, bool):
+            raise TypeError("sort_keys needs to be of type bool!")
+        if not (indent is None or isinstance(indent, int)):
+            raise TypeError("indent needs to be of type integer or None!")
+
+        if export_type == ExportTypes.FILE:
+            if not target:
+                raise ValueError("Please provide a path if your want to export to a file!")
+            with open(target, "w") as f:
+                f.write(self.__prepare_export_output(sort_keys, indent))
+        elif export_type == ExportTypes.STRING:
+            return self.__prepare_export_output(sort_keys, indent)
+        else:
+            raise ValueError("Please use on of the two given options for the export type!")
+
+    def jsontomodels(self):
+        """
+        Convert the loaded JSON to the internal modules. Without calling this the loaded data will not be available for
+        manipulation.
+        """
+        breeds = self.signaturesjson.get(self._rootkey, -1)
+        if breeds == -1:
+            raise AttributeError("Missing Rootkey \"" + self._rootkey + "\".")
+        for key in breeds:
+            breed = OsBreed(key)
+            breed.decode(breeds[key])
+            self.osbreeds.append(breed)
+
+    def addosbreed(self, name: str):
+        """
+        Add a new :class:`OsBreed`.
+
+        :param name: The name of the new breed. Must not exist in the currently loaded models.
+        """
+        if name in [x.name for x in self.osbreeds]:
+            raise ValueError("Breed \"%s\" already in the list of breeds!" % name)
```

Comment:
Would it make sense to log an error and just ignore duplicate breeds?

Reply:
Well this is a library, so I am not sure that logging is really the right thing to do here. In an application I would agree fully but otherwise I would just let the application then handle this error and log it or do whatever it desires to do.